### DISCUSSION TITLE
Introducing the "module" feature

### DISCUSSION THREAD

**Discussion by ricktu288** [ricktu288/ray-optics#145] [Announcements]
![module-example](https://github.com/ricktu288/ray-optics/assets/1409171/073498df-8615-4212-801f-8c01f2784997)

The "module" is a new experimental feature that allows you to combine several objects (optical elements, etc.) into a single reusable "module object" with custom parameters and control points. The objects within a module are defined with a template, which allows math equations and the creation of arrays of objects.

**Note: You don't need to clone the repo to make custom modules. Everything happens directly within in the web app.**

# Examples

- [Ray relaying](https://phydemo.app/ray-optics/simulator/#XQAAAALzAgAAAAAAAABEqcrGU8hqLFnpmU9ERVKNOyc6-VW14TnhxPDkzS9oA6v6kfjm5-x-h7hVU0PRDFcXYUmQqQOvlKSNuPl4yvJl3_Z3-Oij3pknnT_rM8462ZubaaLa9dIuYC8rYOwwgnHxo1iZS1Kh5a7vqInlw43c39sTsN0hcn4m7gc3acZUUIo3fibsa4T9bC-fMbKyymDO02GCOCIAIAVCNbtfOZxZKQ38PNZWDqnmrYh4zUgj5qIOIYLYoNfkkVQvonJLIWCbnIaNbzGW6qwfV1qXCFGOlZlN-U8ZHw057i2e-UNITwuq9s_pvZY8_Pc4PXngbtqBTN01F9vtQ-0oJwLXkGuF6sG2PzOprSRVTQVkszRxTzaJJTmTh2f54nVz4B1ymUx4824cUslIcwPPn63wE2WUyua49G7mba5A680WmY7RBERLMO3ctwGcodaXZtdUZcdGzASf9tekTy2otIiNSTy-NXxxIXiKLXS70C6fEeUbDj_sXC4_Bhd2K8kBbkAzj_aznJ-C39r-gKBizq97T4-z5nwqssI1CHpJ3ShQdWbsiQ2RQMNa9zJ82t3DWCvAWIFYelQOq41YF1ZY-ZK7JGzAcJLc3Zdm1oVKpoCRrZFkzrCgVe_4HqCmFw5j7LfkVv73flmK) (modified from [here](https://phydemo.app/ray-optics/gallery/ray-relaying))
- [Beam expander](https://phydemo.app/ray-optics/simulator/#XQAAAAJeAwAAAAAAAABEKcrGU8hqLFnpmU9ERVKNO0a9YiTICNtkK--u3fsZ6woiVmuyW6wju_tlcADOd9hbMErd1ki66oPDVnfKDYDLkFTDqtmZsg8kYQtmPO5hF8ZVkQg7xz9emOanmdHRU8BRsuwGoEkzqdap5OC3-bocFeITVEW3SN87pE0KBoAOk9oXDwD_3k0BlxdNVwz9hjp81PzuYBz_Q4mCN_WJjCWfWzhTDD33bKs-77mhVtgSZIIECRFBtcPMMfr1hhi1BTIkJa8lnXMy7bCY_qqTOQZM2Nxf7I_Frbm7V7wQQP_47mTiS2DLoDAzK_eEiMHvWeAE9xPGwy-0c19v0--oVjq8Tl_zZUbnOBZKlCnBQi0kGb-Hvnw71-ZvIZFZj1Y4IHmU3cmZafYu1S-3fny66e0mF2ZzIRQG7s_E81SJxPh7egNs3W7SAqow7uf1psQUmvGSdK-1FcfENlKQIno_Y54WqxvBoVWqERWN-M_fBMTJkVb_afnLtsnxmF2O8maV1gOZnIZRk0A-jRvId_BA5ZyjYDq8MIXTpYR5ssNdo9s_cvZkw7HKLRNteqZ8yK4fYyZYKJP_7gmFAD-zOfTZJweOrPkD_AGvUpPY_RiVSWRW3S7ddzf5nhedmmWZ6C0P3DlpF-aBL05FFqSkWP-Qccno) (modified from [here](https://phydemo.app/ray-optics/gallery/beam-expanders))
- [Chaff countermeasure](https://phydemo.app/ray-optics/simulator/#XQAAAAI6BgAAAAAAAABEqcrGU8hqLFnpmU9ERVKNPCvU2vNexoYqNqBXYfVzBRPAMeZIgUUSDjgSrKA8GKCqxf2WxfU7s3X42dR1Z_XqVc-wM9EQvb69ldFaTxrNVhtoz1adown1KXuVcuW_ROE5MXup7aakcHwxYz-ghT3MhKQWlUiu6AkrA9f8UjlTxhBiemCepxxKK5kc-WJz7YejinlNHCJhxUF9j172CVtvtF4QJ-obmCA-w90zYPVJEEji-sn64suHHKVJftUbwWMfq4OM8g3CBVuiitvQ5io1ltdXcNgRJVwqTQiAxWPghbRo1GsKrac4cdcOwS6sPFmuYJlApWDvuyAxpxmwLkjsmvQiGQhlrUveSh_PkJMIP0xGVYF-lNfvZMTlUWQlD922iWo-xFiGDfb3hPGbFqr7QwTGjPj2VDJzV4rFMjRIJ3uvTk3aGnBnuLZm1zJw-WPI6pDFYSSzbTVw5bW3ZeqbsfT0S4LHfTvvAmGBD2FJOPK252r1vrFeNdJspuT07UDG5wYyZw7k1EUm51hTCJLXDJQvx5qkTnJO9N3Rvq-3n2E5gErydrC4Jppuw1h8Jz8Djw2gZCrlaPt1EdNd9b8pfgux5JLKlmx7V4VEPm5spMMBDcYuJUdDizvmODAC4YBuJvKiShI4OQGJVTNV4-UKb2vQFWSyt8RmD2dJDLtXH1x5lZOsEr_GmI2oo7m5LBF8bJJ8cvsHuxScKM3oiTvQvsp8cXcrOaq1VSwfd9h0f3U0iIG8OPFNh7_8oYfcvTXf7RJl_lZjTMSu7yGlHbBmhW_vNqdPw5TgPx3CDQO8IlxaNxbjl2mZwlp6ipsxRMKuWQzUcHkcGoDH8jL6N42KfsvQz0oWT_L6th0U7RmvAG1feSgBiQJ4UeZrFc3ktDFFsODGmRH1XYy0h8FHKdpRnSu4K_OwGId9DvIpA62Diyp-_z9FvlrhIUJF-Jdy__f690E) (modified from [here](https://phydemo.app/ray-optics/gallery/chaff-countermeasure))
- [Fresnel lens](https://phydemo.app/ray-optics/simulator/#XQAAAAKtAwAAAAAAAABEqcrGU8hqLFnpmU9ERVKNOyZTW3qfgW3tiGaEudNMjXImfLdWp76ITdC3_i6nC7RW7-z16HA_KxUYuzbgYw4FaqlKvH0SD63HcvcXsvtcnzRktm6l7Fjbh_PV-sp6dtqM1qDb5WcBZZBZriReB6Ftylm5oQgxIUh4quOIvmUJfrUFqExLD3esMTDK_fbtaOJeHBAM5o48IaZKxhdKquR9m_ekgwmTeEKP-Q-QaFIzcWeWiuxLZY26U6dgJXobW2XbW61VDZxavd6Ahvlzzb_664Dl_0JqLoWsiEP444QHAHfxCgB3O7RpR9sBxy5xf4SuJmGDj3WhZtkx9VoyVMzpK0yHZcv__Gi9K2GdDqWedIrXp-qQeMYDhuy9zNB-Ju28IZb3bUwuNF-SwUV8X-oeU0fecotK_rCflHf1_0KarCsoZfn-gnTh-83oPKvSonkolI3sZTFcUep5pirHiPykY7N0Wgno7_b4wtM1sYXutSDRcl1IBSM9nCsN7AUo7HRNl4F_kHn9fAvUYQf8gmmp2lIFIi0U1XuYoNzCkp89BYGVFFWJlx7lHshKuMTw0usWaT2d26lhS_zB0T35_xuk2GJqRcZV6wMu57FTgFoCyGIHwyAhoc3ueoaxjD48cdwsYo0Nse_USwB4dedeBv7F6UMS2eSPLInPnrCHhcrFi2HfWQRwJX-ATrPuQDFhyP1K68FSZacGOXGskhrvYuL3UNmeJSchUp9G9OWLIq2AsXwlmVj6o2Hdt0anTzGG_t-wgciqUTMldgABTLKKQqA5nBcIZhUeYPToOjRnTrCcLdg30IFybcbQsss8I_seuJY) (modified from [here](https://phydemo.app/ray-optics/gallery/fresnel-lens))
- [Continuous spectrum](https://phydemo.app/ray-optics/simulator/#XQAAAAKHAwAAAAAAAABEKcrGU8hqLFnpmU9ERVKNPAskKpscaTaRLFvp5cCofuQKm22i0caEhjkR9ddJaMMi0IYOtxjjbjcIVndOYKZ5Gv8jXuuZdPajYRsEEG_Ahs_JB3VmqLzUF-vVCofAEgjyanOvo0_lRwLN3kFxFOXZNFIlxYOPZGXMEm4tpVJw4szozZ-tyQMKeEu5VLnF2bshviM_AcbOD2c40Oi9y3JsAhYZhXglpe3cCX4BBc3iIPkuA440gjbCHjeL8bzmH-LRRF6j-Q0VPch7V30E9g-ZMs4XV-TmfG0cnMILpWw2SCs3TQ1yzj6Ir9b_LF_gz9xzPoKBaEWTpGOnNETzXPp21OzNgPvRwfvmTQKoH4RKOOPC7mCQJThA0VWLG6Wd9dVhlCQbQN2jA9vLyv0CUjaL8p5HtTp-tfNEy5EdYtRrNLfP7kMHThHOPfxLzTAyxmXxTbTkj84_gnB2K7GG1Wn9eJ_qTz5nGwER2V3fhtYE8IrsMyMPmyqOc5dwEY56oK75EnmFsfZOMwsrw369oKsYDiloXb1Mtp2OEPAXLeKaJtUM-Ix3gyipZYzmlz9Ni8qCp2OkD5UUsK4TBNuK6fD2f8dA9624CEcZ80fk28z31njNCFfr891p8lGAkJ-Fw_kMpLnQnaxnPwZdD7LuFoz3Y8DEXYwgnJr5S7vVw6_c6ylI8NpLxpPv_cFicFEDV5-pB6dfMF3WB6k33CnE2_ZHccKd8JvKNIV03Oe-nCzM13ZgW0HZ4Uyn_6JsN8s)
- [Circular source](https://phydemo.app/ray-optics/simulator/#XQAAAAKYAgAAAAAAAABEqcrGU8hqLFnpmU9ERVKNO4hA5p7_1EC4mPvg7YZo8XAZLOCF_px6RthCmFTzeJJZYLqokselbnf4jeO_5uXeKQa1jFPSkhT_0kpZKe28sT-KC1P0zG0a392x1BZuiz9m4U9Fua9MJ1WJEqtKOEkeVGlgDPWbRhlm9GJ5Zb9ZqN0WLjStImibrOK5cLeWwus40Zm3qUHi3GDXk1Sry2ZsClbwfkj2QR80UFmvThtycI2tljfuIa7usC11o8NfQVwimiHj-AwJalwwg-7-A0jL4qUDmjRLDaLrOIZCRuxyljoC4tg5Hqmn4A0RYlog2O_W7J6jdlgXzo-gh6JfEc-IMX3mTHOQVOgdp4_3rb_rJtESadxAQ5kpy1i7nDVEK6hdmlVyT2dT0n3AOpwUzVRmv5LVFqXrTdwY36VsOR3OPACrhtceSj0zFjNHPkeJBxlfR4Mb3Ew8fM2WABJGBJBxrjkSe7cOwv_kKOhBDVT2Du7PIe1sl3GEP48brccnIEQf7mRN8JqhXtO2OJ-cAoFHurxU2Mr4lmZR6HtXCepqofYscAXvL7yOyY7-yA3bLDeuOTXekwztTp5Pv4I-gVB2a4LGVZUSwdtwzrTsxP_ffc4F)

_Note: This feature is still experimental, so the corresponding items in the Gallery will not be modified until this feature becomes stable._

# Tutorial

This tutorial will guide you through the basics of creating a module. It works completely within the web app online, so you don't need to install any development environment.

Note that not all custom control points requires a module. Some simple cases may be achieved by the "handle" feature (see "Group, rotate, and scale objects" section in the help popup at the bottom right corner of the simulator). Since creating a module is much more complicated than using the handle, you should check if your case can be achieved by the "handle" feature before considering building a module. See [here](https://phydemo.app/ray-optics/gallery/black-cat-becomes-white) for a non-trivial example of a custom control point (moving two plastic bags out of water) without using module.

## The built-in JSON editor 

This app currently does not have a visual interface for creating modules, so you need to directly edit the JSON of the scene.

You can enable the built-in JSON editor by clicking the "settings" dropdown at the top-right corner of the app, and then check "Show JSON editor". The code editor should appear at the left-hand side of the app, with the JSON code of the current scene. Make sure you have a large enough screen, as this feature does not work well on mobile devices.

As you edit the scene using the usual visual scene editor, the code in the JSON editor will update accordingly, with the changed part highlighted. Conversely, directly editing the code in the JSON editor will update the scene accordingly. If you are not familiar with JSON or any kind of text-based data format, you may wish to play around with it for a while.

In particular, when you add an object to the scene, it is added to the `objs` array. And if you modify some of its properties to a non-default value, they appear as key-value pairs in that object.

## Basics of a module 

Let's look at our first example of a module.

[Example 1](https://phydemo.app/ray-optics/simulator/#XQAAAAKaAQAAAAAAAABDKcrGU8hqLFnpmU9ERVKNOsQbH41v8BdZNCQwRaddFCZLwYsr_x975ihI6J5x23w7otJuIKxwS-4jklMnjhQuuYbgGBgsxu_slLzINF6xmDJ8mbzNwRnRUh5GYwDACulYETTDsYnn-ZDRZ3xiTQsJRvDc0cr1soSQo_kMYOHGa0oUsc64nlfcrvPV0saazctBeq5e2eZwdYYhowFUN3xFIiA8hRJq8APGFONI3RxTCgJH5SDMCJgmhYnF1Sq78y2VuVLErf704tI2)

You should see four lines of texts. By looking at the JSON editor, you will see that the first two are directly in the top-level `objs` array as usual, but the last two are in `modules.ExampleModule.objs` instead. 

The `module` is a dictionary where the key is the name of the module (in this case `ExampleModule`), and the value is the definition of that module. In particular, the `modules.ExampleModule.objs` array describes the (template of) objects within that module, which is different from the top-level `objs` which describes the objects in the scene.

To put the objects within the module to the scene, we need a "module object" in the top-level `objs` array, which is `objs[2]` in this example, whose type is `ModuleObj` and whose `module` property is the name of the module.

The module definition in the `modules` dictionary is not editable by the visual scene editor. So when you click any of the last two texts in this example, you are just selecting the module object, and not the objects in the module. Since the coordinates of the texts in the module definition in this example are absolute coordinates, the last two texts are not draggable. We will learn how to make them draggable by using control points later.

If you select a module object, there is a "Demodulize" button on the object bar. Clicking it will "expand" the module object into its constituent, and `objs` will now contain all the four texts. This operation is not reversible (but of course you can click "undo").

The suggested way of creating a module currently is to first create an empty module using the JSON editor, create some objects using the visual scene editor, and then cut and paste the objects from `objs` to `modules.ModuleName.objs` using the JSON editor.

## Adding parameters 

The objects within the module can be defined by a set of parameters. Let’s look at a simple example:

[Example 2](https://phydemo.app/ray-optics/simulator/#XQAAAALKAQAAAAAAAABDKcrGU8hqLFnpmU9ERVKNOsQbH41v8BdZNDSJRaddFCZLwYsr_x975ihI6J5x23w7otJuIKxwS-4jklMpUeDraVKtJPRrqUDpwm04g0AMDNLFqDuFz-Zv7rT59ZvfcfKcEVFT_07LZ5e_dTZN7hFKeAxYWBohMvZv3d0gGOgqJFXszpCr85ety_ui3xWraqmLTt_qoUyL6GWFQLUrA8q2flAkMhxIOw1D6obib4ifzn_AcKb_26sf5phgyrzWPNWLriSyt7KbW8KcXvKMaRm-W1HXD8uVN3pc5wml8XJoQMRFmlGvYJzLzAlbkW2veqCagQ3CxoR-uQ7PsPFEn9bXS6G0dvYz7ahVL5p_PE6G7p5K1b-ad8n0A_ZQeVeG003rpj73X87VWZzgdjrGSNyyXTfqKrnkhYNi2Akxh2jWiLYnMaiSSkmYRJ0f__-CT5kA)

Here `modules.ModuleName.params` is an array of strings `"name=start:step:end:default"` defining the name of the variables and the range of the sliders. The sliders appear on the object bar when the module object is selected.

Within the `modules.ExampleModule.objs` array, any values can be expressed using those parameters. Within a string (such as the `text` property of a `TextLabel`), the equations of the variables are enclosed by a pair of backticks. For number parameters (such as the `fontSize` property of a `TextLabel`), you need to make it a string so that you can use the backtick format in it, so each equation is sandwiched by a pair of backticks and a pair of quotes. The equation are evaluated with math.js (https://mathjs.org/docs/reference/functions/evaluate.html). See there for the available syntax and functions you can use in the equations.

The actual values of the parameters are stored in the `params` property of the module object, which, unlike the module definition, can be directly edited by the scene editor using the slider.

## Adding control points

To make the module object draggable, we need to parametrize the objects within the module using a set of control points. Let’s look at the example:

[Example 3](https://phydemo.app/ray-optics/simulator/#XQAAAALCAQAAAAAAAABDKcrGU8hqLFnpmU9ERVKNOsQbH41v8BdZNETiRaddFCZLwYsr_x975ihI6J5x23w7otJuIODVK-4jklMnjhQuuYbj8KAsxu_slLzINF6xmDJ8mbzF5b2pG646ya9uBnARPzsT-OTOuapzhLdnV8lt-x_VeyB6I_QliDleHgiquk2cl1loiDOYRaYBWW3evJsWkDgNp00dGYGpvOeGfYnufZXZrrQLNz9pBMthPalWj_4pVQdcLgPiLj076lhejKTd8G7n24yBA1BGHlkkblUivDd27ZoiPG5sDKSipUg87Z41jxvbROrD4ohW7x5A1nlmt_7ukT3PXgQ2H6fvJdPuKv1YiRg)

Here `modules.ModuleName.numPoints` defines the number of control points. The coordinates of the control points are (`x_1`, `y_1`), (`x_2`, `y_2`), etc, and are used in the same ways as the parameters within `modules.ExampleModule.objs` as described by the previous section. Note that the index starts from 1.

The actual values of the coordinates of the control points are stored in the `points` property of the module object, which, unlike the hard-coded coordinates in Example 1, can be edited by the visual scene editor by dragging the control points, each shown as two concentric gray circles in the scene. If you drag elsewhere in the module object (such as dragging the text labels), all the control points will move together.

Since our module object can now move, it is now quite easy to create multiple instances as in usual tools. The name of the module is shown in the Tools -> Other menu, and you can select that and then click two points in the blank space in sequence for the two control points to create another instance of the module.  You can also use the “duplicate” button on the object bar.

## Arrays and conditionals

More complicated module can be built using arrays and conditionals. Let’s look at the example:

[Example 4](https://phydemo.app/ray-optics/simulator/#XQAAAAK8AQAAAAAAAABDKcrGU8hqLFnpmU9ERVKNOsQbH41v8BdZNFU7RaddFCZLwYsr_x975ihI6J5x23w7otJuIKxwS-4jklMocmzraU-uionNR68scp3EjMbxo075DzkS5bn92aCj7fMpqtcEQGyOIa3Ytn5fUKD0CpLWznIyPPOiLWhiwOmcPQlwan-fs70FX-DM6SgNNF2EjuNZZo1y2waCYKGbdhxs-7am-OdsuyXPd17KrS7d-SQU0Dm_8VkAiWS4HDNUolMU9wdOoVQq29HyhfYjOTbeSV1VjosyVzdoh7amq0ti7gldkuzrSP9f1u3Fk4PLRopc41C8M6Q97E9uU5bLkzeclJQuATnfoii0luX17mXaxCRRq7Rxl1O5KhcbCPxqjYNugZQ6Ax5U6C5O74wvONieSoSs1tnYo1iDIFbFqsciQnaZ9jfFk__N-Oys)

Within `modules.ExampleModule.objs`, any objects in an array can have two special keys: `"for"` and `"if"`. The value of the `"for"` key is either a string of the format `"name=start:step:end"` defining a loop variable, or an array of several strings of this format describing a multidimensional loop. Such an object in the array is duplicated several times according to the loop variables. The value of the `"if"` key is a string representing a math.js expression that evaluates to a boolean, and such an object is included in the array if and only if the boolean is true.

To prevent accidental infinite loop, the total number of iteration of each `"for"` loop is limited by the `maxLoopLength` property of the module definition, whose default value is 1000. You can set this property to a larger value if needed.

## Objects with built-in custom equations

For objects that already have custom equation input (such as Mirror -> Custom Equation), the equation property in the JSON is a string representing a LaTeX equation, rather than a math.js expression. To include custom parameters in the equation, you must use the same template syntax as if the LaTeX equation were a regular text. So the part enclosed by the backticks is in math.js expression, while the part outside is in LaTeX. The module parameters can only be accessed in the math.js part, and the independent variables of the custom equation (e.g. $x$) can only be accessed in the LaTeX part. Here is an example of generating a mirror with equation $y=\cos(2\pi x+\phi)$, where $\phi$ is a module parameter:

[Example 5](https://phydemo.app/ray-optics/simulator/#XQAAAAJlAQAAAAAAAABDKcrGU8hqLFnpmU9ERVKNOsQbH41v8BdZNGWURaddFCZLwYsr_x975ihI6J5x23w7otJuIKxwS-4jklMod0cmCJaNLvHXaJqidQycgLCxvo_MUR4pKSF02EThSWEmzI1Nvg8p3rGzbgOCoW20RA6HEvR3WRy-n0CVOj501Zey96roKU99sYBaUujCSyd7TtjFHhGSTNA9sDcUsOu9tx7I6VVyQkECCxRMQp8unR3jEwejYdZwvzxqR8UFHrvfH4S8g0eb5UpBI_n2Wdm84RwKJ2haZ53zK4phO0cPMI6j41Zj0hPk1iWMWI15L1Li6CURx1HXAcQcPRZo8rlvz0965RPAe0gsbc4XegcJboQnpvSzlz__djf1AA)

In the future, there may be a way to unified the equation input.

## Objects with built-in shape parametrization
For objects that already support different ways to define its shape (currently only Glass -> Spherical lens). There are special JSON syntax for such objects that can be used within the module definition, even if they are always defined by shape in the top level `objs` array. Here is an example:

[Example 6](https://phydemo.app/ray-optics/simulator/#XQAAAALXAQAAAAAAAABDKcrGU8hqLFnpmU9ERVKNOsQbH41v8BdZNHXtRaddFCZLwYsr_x975ihI6J5x23w7otJuIKxwS-4jklMnjhQuuYbgHTx4xu_svNmoEFyZi2EAGHhnylj8kZWffIJDQr5n_mkzRL8qYAyDV2doAoIpGCABFw4k4ofmmnIwAWmlAgPZSMvmxNhjgk_S8gXQf3VE6ScOowAVIe31VTcW1pGd6WByGgtLfSY4fiv0lNdI-HjIDZ68rY_xdq4-X68TXz3o1dfMgaJSfniRIpqkDJXkfWiDVxtWxp7WQcahm_-cjgLQf9Hn3PXaleaqP-s8YmdqVK1WG7wL_pWnuNUOtHZgD4IhnzGd4LLm8VK9DRfOxigrQtwGyjpOEnsEpoOeE_7noiQ)

# Contributing

## Contributing a module
You are welcome to contribute to the list in Tools -> Other -> Import Modules. Please add items to the [`modules` folder](https://github.com/ricktu288/ray-optics/tree/master/modules) and open a PR, or send the file to ray-optics@phydemo.app if you are not familiar with GitHub.

## Expanding the capability of modules
See [`ModuleObj.js`](https://github.com/ricktu288/ray-optics/blob/master/simulator/js/objs/special/ModuleObj.js) for the source code. See #142 and #143 for currently required capabilities which are still not directly possible.

_Note: The `modulize` branch is not related to this feature; the "module" there refers to the different concept of (slowly) refactoring the code of the app itself towards a modular structure._

## Possible visual editor?
A graphical editor of modules may be desired for the future, but it will be quite complicated. See https://github.com/ricktu288/ray-optics/discussions/141#discussioncomment-9133179 for a possible approach.

**Comment by ChenZhu-Xie** [comment]
When an object is edited, the corresponding embedded JS code editor immediately shows the updated code for the altered object. 

I appreciate this "display corresponding source code" feature, as it is similar to LaTeX's click-to-source functionality in PDFs. 

I'm curious about how this can be implemented with JS :)

**Reply by ricktu288** [reply]
The current approach is to simply take the difference between the old and new JSON string, get the position of the first and the last characters which are different, and select between them. This is not real synchronization like SyncTeX.

I think the proper approach should be something like https://github.com/epoberezkin/json-source-map

**Reply by ricktu288** [reply]
If we want that when the user clicks a content object of a module object, then JSON editor will jump to the corresponding content object in the module definition, then the proper synchronization will be needed. But I think it will be quite complicated to implement.

**Comment by brainandforce** [comment]
How do you embed a module within a module?

**Reply by ricktu288** [reply]
All modules (parent, child) must be defined directly in the "modules" object, then in the "objs" array of the definition of parent module, put one or more "ModuleObj" which are instances of the child module.

The "Coated lens" and "Chaff" modules in the Import modules list are such examples
